src/pages/HabitTracker/Habits.tsx

```diff
@@ -1,87 +1,256 @@
-import "bootstrap/dist/css/bootstrap.min.css";
 import { useState, useEffect } from "react";
+import styled, { keyframes } from "styled-components";
 import { Modal, Button, Form } from "react-bootstrap";
-import styled from "styled-components";
 import _ from "lodash";
 import { PageBackground } from "../../components/styled/PageBackground";
 import { Sidebar } from "../../components/Sidebar";
 import { links } from "../../common/SidebarLinks";
 import { HabitTile } from "../../components/habittracker/HabitTile";
-import { AddHabitTile } from "../../components/habittracker/AddHabitTile";
+import { AddHabitButton } from "../../components/habittracker/AddHabitButton";
 import { StyledModal } from "../../components/styled/Modal";
+import { EditModeButton } from "../../components/habittracker/EditModeButton";
+import { DateRangePicker } from "../../components/habittracker/DateRangePicker";
+import { LogButton } from "../../components/habittracker/LogButton";
+import { EditButton } from "../../components/habittracker/EditButton";
+import { DeleteButton } from "../../components/habittracker/DeleteButton";
+import { RemoveLogButton } from "../../components/habittracker/RemoveLogButton";
+
+// Utility function to update habit data in state and localStorage
+const updateHabits = (
+  updatedHabits: Habit[],
+  setHabits: (habits: Habit[]) => void
+) => {
+  setHabits(updatedHabits);
+  localStorage.setItem("habits", JSON.stringify(updatedHabits));
+};
+
+// Utility function to update logging data in state and localStorage
+const updateLogging = (
+  updatedLogging: any,
+  setLogging: (logging: any) => void
+) => {
+  setLogging(updatedLogging);
+  localStorage.setItem("logging", JSON.stringify(updatedLogging));
+};
+
+// Utility function to reset habit form
+const resetHabitForm = (
+  setNewHabit: (habit: Habit) => void,
+  setOriginalHabit: (habit: Habit) => void
+) => {
+  setNewHabit({ name: "", count: 0 });
+  setOriginalHabit({ name: "", count: 0 });
+};
+
+// Utility function to manage logging data
+const manageLogging = (
+  habitName: string,
+  selectedDate: Date,
+  logging: any,
+  setLogging: (logging: any) => void,
+  action: "add" | "remove"
+) => {
+  const year = selectedDate.getFullYear();
+  const month = selectedDate
+    .toLocaleString("default", { month: "long" })
+    .toLowerCase();
+  const day = selectedDate.getDate();
+
+  const updatedLogging = { ...logging };
```

Comment:
I like what you have done here. I hope you understand why we have to spread this object. If you don't understand this concept, ask me next time we have a meeting.

Spread operator copies the contents of the object if they are nested only up to 1 level. Rest all of the nestings are used as a reference. I'll explain it to you in detail during the meeting.

Best option here is to use the deep clone from lodash.